atlasdb-commons/src/main/java/com/palantir/common/concurrent/ConcurrentStreams.java

```diff
@@ -50,33 +53,34 @@ private ConcurrentStreams() {}
             return values.stream().map(mapper);
         }
 
-        List<CompletableFuture<S>> futures = Lists.newArrayListWithCapacity(size);
-        for (int i = 0; i < size; i++) {
-            futures.add(new CompletableFuture<>());
-        }
+        Map<T, CompletableFuture<S>> futuresByValue = values.stream()
+                .collect(Collectors.toMap(Function.identity(), value -> new CompletableFuture<>()));
+        Queue<T> valuesToProcess = new ConcurrentLinkedQueue<T>(values);
 
         if (size < concurrency) {
             concurrency = size;
         }
-        AtomicInteger nextValueToProcess = new AtomicInteger(0);
         for (int i = 0; i < concurrency; i++) {
             executor.execute(() -> {
-                int currentValueToProcess;
-                while ((currentValueToProcess = nextValueToProcess.getAndIncrement()) < size) {
-                    T value = values.get(currentValueToProcess);
-                    CompletableFuture<S> future = futures.get(currentValueToProcess);
+                while (!valuesToProcess.isEmpty()) {
+                    T value = valuesToProcess.poll();
+                    if (value == null) {
```

Comment:
nit: don't need this due to the !empty check

Reply:
What if two threads try to poll on the queue when it has one element left?